### PR TITLE
Rag evaluation article - part 1 

### PR DESCRIPTION
**Pull request by atarora**

### CODE REVIEW

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+### Let's dive in !!
+
+## Strategies for Evaluation
+
+Now that we have defined different challenges and levels at which we can break down the RAG evaluation, let us zoom into the levels individually. 
+
+### Model Evaluation 
+
+To evaluate our model, we start with this [Massive Text Embedding Benchmark](https://github.com/embeddings-benchmark/mteb#leaderboard). We want to ensure that
```
Comment (kacperlukawski) [inline docs/use_cases/retrieval_augmented_generation_eval.md:83]: How about putting a direct link to the leaderboard?
https://huggingface.co/spaces/mteb/leaderboard

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+print("--DONE--")
+```
+
+### Data Ingestion Evaluation
+
+We first validate the model, and train it on the language of our domain. 
+We can then configure data ingestion into our semantic retrieval store aka vector store.
+Various vector databases offer index configurations to influence and enhance the retrieval quality, based on the supported index types (Flat , LSH , HNSW and IVF).
+One such example here to improve HNSW [retrieval-quality]("https://qdrant.tech/documentation/tutorials/retrieval-quality/").
```
Comment (kacperlukawski) [inline docs/use_cases/retrieval_augmented_generation_eval.md:114]: The link is not rendered correctly in the GitHub preview. Maybe the quotation marks should not be put around the URL?

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+To evaluate ingestion, we need to focus on related variables, such as:
+
+* **Chunk size** - represents the size of each segment. This depends on the token limit of our embedding model. It also has a substantial impact on the contextual understanding of the data. That impacts the precision, recall, and relevancy of our results.
+
+* **Chunk overlap** - represents the presence of overlapping information fragments in each segment. This helps with context retention of the information chunks but at the same time must be used with relevant strategies like deduplication, and content normalization to eradicate adverse effects.
+
+* **Chunking/Text splitting strategy** - represents the process of data splitting and further treatment as mentioned above.
+
+A [utility]("https://chunkviz.up.railway.app/") like this seem useful to visualise your apparent chunks.
```
Comment (kacperlukawski) [inline docs/use_cases/retrieval_augmented_generation_eval.md:124]: Again, the link doesn't render on GH preview. 

```suggestion
A [utility]("https://chunkviz.up.railway.app/") like this seems useful to visualise your apparent chunks.
```

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+It provides a common ground for objective comparison and improvement of a retrieval process/algorithm.
+
+
+### End-to-End Evaluation
+
+An end-to-end evaluation covers the response generation of the question. It leverages the provided context through retrieved documents.
+
+Evaluating the quality of responses produced by large language models can be a challenge due to various factors as described above. 
+
+By virtue of nature and design , the answers generated rely on diversity of response which makes it impossible to device a fixed metric or methodology that fits in all domains and use-cases.
```
Comment (kacperlukawski) [inline docs/use_cases/retrieval_augmented_generation_eval.md:157]:
```suggestion
By virtue of nature and design, the answers generated rely on a diversity of response which makes it impossible to devise a fixed metric or methodology that fits all domains and use-cases.
```

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+For more information, see this paper which provides great ideas to establish [Quality Criteria]("https://scholarspace.manoa.hawaii.edu/server/api/core/bitstreams/c6a53998-09e3-4d17-91fd-c7416d51b250/content") for the same.
+
+To summarize, you want to establish methods to automate evaluating similarity
+and content overlap between generated response and reference summaries. You can
+also leverage human evaluation to evaluate subjective information, such as context-relevance, novelty, and fluency.
+
+You can also build a classified 'domain - questions' set based on question complexity (easy, medium, hard). It can give you an overall sense of the RAG performance.
+
+Nevertheless, it is a challenge to formulate a comprehensive set of metrics. It
+has to appraise he quality of responses from LLMs. It remains an ongoing and intricate issue in the context of natural language processing.
```
Comment (kacperlukawski) [inline docs/use_cases/retrieval_augmented_generation_eval.md:174]:
```suggestion
Nevertheless, it is a challenge to formulate a comprehensive set of metrics. It
has to appraise the quality of responses from LLMs. It remains an ongoing and intricate issue in the context of natural language processing.
```

**Review by kacperlukawski — CHANGES_REQUESTED** [review]
There is a general issue with rendering the links. Just a few of them are really visible in the rendered Markdown: https://github.com/superlinked/VectorHub/blob/46ccd8ccf92741454fc6d3772cc7edef027de8cc/docs/use_cases/retrieval_augmented_generation_eval.md

I think it's mostly due to using quotation marks around the URLs, and should be fixed by removing them.

Overall, I made some minor comments only.

File: docs/assets/use_cases/retrieval_augmented_generation_eval/rag_challenges.png
[the remark this answers is not in the document]
Comment (atarora) [reply docs/assets/use_cases/retrieval_augmented_generation_eval/rag_challenges.png:1]: It is one of the slide from the RAG eval presentation deck , let me add that as a link under the image , or would you suggest something else?

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+>If you can't retrieve it, you can't generate it!!
+
+Our experience has taught us that it is better to evaluate each component in isolation.
+
+As for different components of RAG viz. Information Retrieval - Context Augmentation - Response Generation , we classified challenges of RAG as below :
+
+<figure>
+<img src=/assets/use_cases/retrieval_augmented_generation_eval/rag_challenges.jpg alt="Classification of Challenges of RAG Evaluation" data-size="100" />
+<figcaption align="center">Slide from RAG Evaluation Presentation mentions ['Lost in the Middle' problem](https://arxiv.org/abs/2307.03172)</figcaption>
```
Comment (kacperlukawski) [inline docs/use_cases/retrieval_augmented_generation_eval.md:62]: I think it's impossible to put a markdown link in an HTML snippet as it's not rendered properly.

```suggestion
<figcaption align="center">Slide from RAG Evaluation Presentation mentions <a href="https://arxiv.org/abs/2307.03172">'Lost in the Middle' problem</a></figcaption>
```

**Review by kacperlukawski — APPROVED** [review]

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+> If you can't quantify it, you can't improve it!!
+
+For RAG it can be restated as :
+
+>If you can't retrieve it, you can't generate it!!
+
+Our experience has taught us that it is better to evaluate each component in isolation.
+
+As for different components of RAG viz. Information Retrieval - Context Augmentation - Response Generation , we classified challenges of RAG as below :
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:58]: is there parentheses missing?

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+## Strategies for Evaluation
+
+Now that we have defined different challenges and levels at which we can break down the RAG evaluation, let us zoom into the levels individually. 
+
+### Model Evaluation 
+
+To evaluate our model, we start with this [Massive Text Embedding Benchmark](https://huggingface.co/spaces/mteb/leaderboard). 
+We want to ensure that the model can understand the data that we encode.
+The benchmark shown above leverages different public/private datasets to evaluate and report on the different capabilities of individual models.
+If you're working with specialized domains, you may want to put together a specialized dataset to teach the model.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:88]:
```suggestion
If you're working with specialized domains, you may want to put together a specialized dataset to train the model.
```

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+print("--DONE--")
+```
+
+### Data Ingestion Evaluation
+
+We first validate the model, and train it on the language of our domain. 
+We can then configure data ingestion into our semantic retrieval store aka vector store.
+Various vector databases offer index configurations to influence and enhance the retrieval quality, based on the supported index types (Flat , LSH , HNSW and IVF).
+One such example here to improve HNSW [retrieval-quality](https://qdrant.tech/documentation/tutorials/retrieval-quality/).
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:115]: flagging this to point to building blocks @robertdhayanturner

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+* **Chunking/Text splitting strategy** - represents the process of data splitting and further treatment as mentioned above.
+
+A [utility](https://chunkviz.up.railway.app/) like this seems useful to visualise your apparent chunks.
+
+### Semantic Retrieval Evaluation
+
+The next step is semantic retrieval evaluation. It puts the work you've done so far to a litmus test.
+
+Retrieval is the vital component of the RAG and its evaluation can be addressed as a classic information retrieval evaluation problem.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:131]:
```suggestion
Retrieval is a vital component of RAG and its evaluation can be addressed as a classic information retrieval evaluation problem.
```

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+
+## Understanding Retrieval-Augmented Generation
+
+<img src=/assets/use_cases/retrieval_augmented_generation_eval/rag_qdrant.jpg alt="Implementation of RAG using Qdrant as a vector database" data-size="100" />
+
+**Disclaimer**:
+The provided implementation serves as an example using  [Qdrant](https://qdrant.tech). Alternative Vector Databases are also available. To determine the most suitable Vector Database for your specific use case, please refer to the [Vector DB feature matrix](https://vdbs.superlinked.com/)
+
+**RAG** stands for **Retrieval Augmented Generation**, and it probably is the most useful application of large language models lately.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:14]: maybe link to building blocks instead of explaining RAG in the article? @robertdhayanturner

**Review by morkapronczay — APPROVED** [review]
some minor suggestions and flags for Robert for style review.

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+Its capacity to amalgamate specific, relevant details from multiple sources and generate accurate and relevant content has a lot of potential in various domains like content creation, question & answer application, and information synthesis.
+
+Read more: [here](https://hub.superlinked.com/retrieval-augmented-generation)
+
+## Why do we need RAG?
+
+RAG plays a crucial role to significantly enhance [vector search](https://qdrant.tech/documentation/overview/vector-search/) with the power of Large Language Model (LLM), enabling dynamic content generation based on the retrieved knowledge.
+
+It proves invaluable for tasks requiring detailed, coherent explanations, summaries, or responses that transcend the explicit data stored in vectors. 
+While vector search efficiently finds similar items, RAG takes a step further, offering content synthesis and a deeper level of understanding.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:27]: I would also highlight that compared to just retrieving chunks and presenting them to the user, every answer can be unique and tailored to the exact query. Maybe it is there implicitly but I would I think state that explicitly.

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 7 above)
Comment (atarora) [reply docs/use_cases/retrieval_augmented_generation_eval.md:58]: It would help if you could add a suggestion, may be?

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+model = SentenceTransformer(model_name)
+evaluation = MTEB(task_langs=["en"])
+evaluation.run(model, output_folder=f"results/{model_name}", eval_splits=["test"])
+
+print("--DONE--")
+```
+
+### Data Ingestion Evaluation
+
+We first validate the model, and train it on the language of our domain.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:112]: This is basically the previous section, I don't really see the purpose of this sentence here like this. I'd rephrase this.

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 9 above)
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:115]: HNSW is improving retrieval quality? It rather makes it quick for a very little performance cost, as far as I know. Or do I misunderstand the sentence?

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+### Data Ingestion Evaluation
+
+We first validate the model, and train it on the language of our domain. 
+We can then configure data ingestion into our semantic retrieval store aka vector store.
+Various vector databases offer index configurations to influence and enhance the retrieval quality, based on the supported index types (Flat , LSH , HNSW and IVF).
+One such example here to improve HNSW [retrieval-quality](https://qdrant.tech/documentation/tutorials/retrieval-quality/).
+
+To evaluate ingestion, we need to focus on related variables, such as:
+
+* **Chunk size** - represents the size of each segment. This depends on the token limit of our embedding model. It also has a substantial impact on the contextual understanding of the data. That impacts the precision, recall, and relevancy of our results.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:119]: It also controls the granularity of the information in the chunks. Depends a lot on what type of queries we expect.

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+* **Chunk overlap** - represents the presence of overlapping information fragments in each segment. This helps with context retention of the information chunks but at the same time must be used with relevant strategies like deduplication, and content normalization to eradicate adverse effects.
+
+* **Chunking/Text splitting strategy** - represents the process of data splitting and further treatment as mentioned above.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:123]: how is this different from chunk_size and chunk_overlap? I suppose you mean what constitute the base splits that are later merged into chunks, right?

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+us to identify reference metrics and important parameters. Done correctly, it
+helps us determine if the documents retrieved at this stage are relevant
+results.
+
+We have several existing metrics to guide and define our baseline:
+
+- Precision and Recall or their combination F1 Score
+- DCG and nDCG, which relates to their relevance, based on the inclusion or rank of documents in the results
+
+The nature of semantic information retrieval poses a challenge at this stage, as the documents are retrieved beyond the keywords/synonyms/token enrichment-matching.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:143]: I don't undestand this sentence, I guess readers won't either

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+Evaluating the quality of responses produced by large language models can be a challenge due to various factors as described above. 
+
+By virtue of nature and design, the answers generated rely on the diversity of response which makes it impossible to devise a fixed metric or methodology that fits all domains and use-cases.
+
+To address these difficulties, you may use a blend of existing metrics like the following scores: 
+
+- [BLEU](https://huggingface.co/spaces/evaluate-metric/bleu) 
+- [ROUGE](https://huggingface.co/spaces/evaluate-metric/rouge) 
+
+You can then combine these score with LLM-based or human evaluation methods.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:165]: I think here it probably could be explained a little more that the best and kinda online really good evaluation method here is a human deciding whether the answer is good or not. That can be outsourced to a good quality LLM, like GPT-4.

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+For more information, see this paper which provides great ideas to establish [Quality Criteria](https://scholarspace.manoa.hawaii.edu/server/api/core/bitstreams/c6a53998-09e3-4d17-91fd-c7416d51b250/content) for the same.
+
+To summarize, you want to establish methods to automate evaluating similarity
+and content overlap between generated response and reference summaries. You can
+also leverage human evaluation to evaluate subjective information, such as context-relevance, novelty, and fluency.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:170]: ahh, it is explained here, I would somehow rework this paragraph and the previous one together

**Review by morkapronczay — COMMENTED** [review]
Left some more, a bit more substance matter related comments, feel free to disagree or start a discussion about them! :)

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 12 above)
Comment (atarora) [reply docs/use_cases/retrieval_augmented_generation_eval.md]: Done! Please review

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+
+For evaluation, we need metrics/methodology/frameworks that uses one or more
+tools. The tools that you use should cover each component of RAG. Such coverage
+ensures granular and thorough measurements.
+
+Evaluation metrics can assess:
+
+- Retrieval effectiveness
+- Coherence of generated responses
+- Relevance to the retrieved information.
```
[the remark this answers is not in the document]
Comment (atarora) [reply docs/use_cases/retrieval_augmented_generation_eval.md:77]: I can think of rephrasing it if it isn't clear but what I meant to assert : 
Effectiveness being measure of efficiency and accuracy (recall)
and relevance being measure of meaningfulness and appropriateness wrt to query (precision / F1). 
I agree they are closely connected !

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 15 above)
Comment (atarora) [reply docs/use_cases/retrieval_augmented_generation_eval.md:127]: This was the basic idea behind it : https://js.langchain.com/docs/modules/data_connection/document_transformers/#types-of-text-splitters

Do you suggest I explicitly add a couple of examples to facilitate understanding?

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 16 above)
Comment (atarora) [reply docs/use_cases/retrieval_augmented_generation_eval.md:147]:
```suggestion
At this stage, the challenge arises in semantic information retrieval, where documents are retrieved based on criteria extending beyond traditional keyword, synonym, or token-level enrichment matching
```

Comment (atarora) [reply docs/use_cases/retrieval_augmented_generation_eval.md:147]: Does this seem better?

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 7 above)
Comment (morkapronczay) [reply docs/use_cases/retrieval_augmented_generation_eval.md:62]: I did not add because I am not sure what this sentence mean. Maybe 
```suggestion
As for different components of RAG (Information Retrieval - Context Augmentation - Response Generation), we classified challenges of RAG as below :
```

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (morkapronczay) [reply docs/use_cases/retrieval_augmented_generation_eval.md:77]: Ahh, I see. That was not clear for me, maybe a bit more explicitly rephrasing that would be great!

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 15 above)
Comment (morkapronczay) [reply docs/use_cases/retrieval_augmented_generation_eval.md:127]: I think I would make it more clear, that there are splits, which are merged to constitute chunks. And there are considerations when creating splits, and also when merging splits to arrive at chunks.

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 16 above)
Comment (morkapronczay) [reply docs/use_cases/retrieval_augmented_generation_eval.md:147]: Still feels a bit convoluted sentence but now I understand that it is about the bag-of-word(token) or semantic vector based representation of text :) This is I think fine by me!

File: docs/use_cases/retrieval_augmented_generation_eval.md (hunk 7 above)
Comment (atarora) [reply docs/use_cases/retrieval_augmented_generation_eval.md:62]: ah viz. is a synonym for 'namely' and hence parenthesis wasn't needed.

Comment (morkapronczay) [reply docs/use_cases/retrieval_augmented_generation_eval.md:62]: I see, didn't know. the hyphen part is still odd but I'll leave this to @robertdhayanturner

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+model = SentenceTransformer(model_name)
+evaluation = MTEB(task_langs=["en"])
+evaluation.run(model, output_folder=f"results/{model_name}", eval_splits=["test"])
+
+print("--DONE--")
+```
+
+### Data Ingestion Evaluation
+
+We first validate the model, and train it on the language of our domain.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:115]:
```suggestion
After we trained our model on the language of our domain,
```

File: docs/use_cases/retrieval_augmented_generation_eval.md [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+evaluation = MTEB(task_langs=["en"])
+evaluation.run(model, output_folder=f"results/{model_name}", eval_splits=["test"])
+
+print("--DONE--")
+```
+
+### Data Ingestion Evaluation
+
+We first validate the model, and train it on the language of our domain. 
+We can then configure data ingestion into our semantic retrieval store aka vector store.
```
Comment (morkapronczay) [inline docs/use_cases/retrieval_augmented_generation_eval.md:116]:
```suggestion
we can then configure data ingestion into our semantic retrieval store aka vector store.
```

**Comment by robertdhayanturner** [comment]
@atarora 
Per our conversation, just two things left to address:

line 16 - your original text here was "The provided implementation serves as an example using  [Qdrant]("https://qdrant.tech")."
It's best to introduce what you are trying to achieve with your implementation before you refer to it. Are you using Qdrant to perform RAG evaluation? Are you referring to an implementation of RAG that you are then going to evaluate afterwards?

line 28 - "In the process of scaling RAG from Proof-of-concept (POC) to production, we learned many things."
Your readers will want to know 1) who the "we" is that is writing this - Are you a team working on RAG applications? I can extrapolate this a bit myself, but it's better if it comes from you; and 2) a few details of your "process of scaling RAG from Proof-of-concept (POC) to production" - whatever you can add without compromising client confidentiality.

Thx!

**Comment by atarora** [comment]
@robertdhayanturner 
To this : 

> line 16 - your original text here was "The provided implementation serves as an example using Qdrant." It's best to introduce what you are trying to achieve with your implementation before you refer to it. Are you using Qdrant to perform RAG evaluation? Are you referring to an implementation of RAG that you are then going to evaluate afterwards?

I wonder : What is unclear from current line 16 that clearly states : Qdrant is used as a knowledge store for our RAG.

**Comment by robertdhayanturner** [comment]
@atarora 
line 16 - what is "the implementation"? i.e., what is your RAG? Are you referring to the evaluation you will perform later? Are you referring to moving from PoC to product for your clients?
